File: android/app/src/main/java/com/deviceguard/core/alarmengine/AlarmEngine.kt
package com.deviceguard.core.alarmengine

import com.deviceguard.core.guardengine.AlarmIntent
import com.deviceguard.core.guardengine.GuardDecision

sealed interface AlarmAction {
    data object None : AlarmAction
    data class Notify(val deviceId: String, val title: String, val body: String) : AlarmAction
    data class Ring(val deviceId: String, val title: String, val body: String) : AlarmAction
    data class Stop(val deviceId: String) : AlarmAction
    data class Pause(val deviceId: String, val untilEpochSeconds: Long) : AlarmAction
}

interface AlarmNotifier {
    fun notify(title: String, body: String, channel: AlarmNotificationChannel)
}

interface AlarmSoundPlayer {
    fun startRinging()
    fun stopRinging()
}

enum class AlarmNotificationChannel {
    PRE_ALERT,
    STRONG_ALARM
}

class AlarmEngine(
    private val notifier: AlarmNotifier,
    private val soundPlayer: AlarmSoundPlayer
) {
    private val activeRingingDeviceIds = mutableSetOf<String>()
    private val pausedUntilByDeviceId = mutableMapOf<String, Long>()

    fun handle(decision: GuardDecision, deviceName: String, nowEpochSeconds: Long): AlarmAction {
        val pausedUntil = pausedUntilByDeviceId[decision.deviceId]
        if (pausedUntil != null && pausedUntil > nowEpochSeconds) {
            return AlarmAction.None
        }

        return when (decision.alarmIntent) {
            AlarmIntent.None -> {
                stopIfNeeded(decision.deviceId)
                AlarmAction.None
            }
            is AlarmIntent.NotifyAndVibrate -> {
                val title = "设备可能离开"
                val body = "$deviceName 可能离开看护范围"
                notifier.notify(title, body, AlarmNotificationChannel.PRE_ALERT)
                AlarmAction.Notify(decision.deviceId, title, body)
            }
            is AlarmIntent.RingAndVibrate -> {
                val title = "设备已离开看护范围"
                val body = "$deviceName 仍未恢复连接"
                notifier.notify(title, body, AlarmNotificationChannel.STRONG_ALARM)
                activeRingingDeviceIds += decision.deviceId
                soundPlayer.startRinging()
                AlarmAction.Ring(decision.deviceId, title, body)
            }
        }
    }

    fun acknowledge(deviceId: String): AlarmAction {
        pausedUntilByDeviceId.remove(deviceId)
        stopIfNeeded(deviceId)
        return AlarmAction.Stop(deviceId)
    }

    fun pause(deviceId: String, nowEpochSeconds: Long, durationSeconds: Long = 600): AlarmAction {
        val until = nowEpochSeconds + durationSeconds
        pausedUntilByDeviceId[deviceId] = until
        stopIfNeeded(deviceId)
        return AlarmAction.Pause(deviceId, until)
    }

    fun pausedUntil(deviceId: String): Long? = pausedUntilByDeviceId[deviceId]

    private fun stopIfNeeded(deviceId: String) {
        val removed = activeRingingDeviceIds.remove(deviceId)
        if (removed && activeRingingDeviceIds.isEmpty()) {
            soundPlayer.stopRinging()
        }
    }
}

class RecordingAlarmNotifier : AlarmNotifier {
    val notifications = mutableListOf<Triple<String, String, AlarmNotificationChannel>>()

    override fun notify(title: String, body: String, channel: AlarmNotificationChannel) {
        notifications += Triple(title, body, channel)
    }
}

class InMemoryAlarmSoundPlayer : AlarmSoundPlayer {
    var isRinging: Boolean = false
        private set

    override fun startRinging() {
        isRinging = true
    }

    override fun stopRinging() {
        isRinging = false
    }
}

File: ios/GuardApp/Core/AlarmEngine/AlarmEngine.swift
import Foundation
import UserNotifications

enum AlarmAction: Equatable {
    case none
    case notify(deviceId: String, title: String, body: String)
    case ring(deviceId: String, title: String, body: String)
    case stop(deviceId: String)
    case pause(deviceId: String, until: Date)
}

protocol AlarmNotifying {
    func notify(title: String, body: String, sound: UNNotificationSound?)
}

protocol AlarmSoundPlaying {
    func startRinging()
    func stopRinging()
}

final class AlarmEngine {
    private let notifier: AlarmNotifying
    private let soundPlayer: AlarmSoundPlaying
    private var activeRingingDeviceIds: Set<String> = []
    private var pausedUntilByDeviceId: [String: Date] = [:]

    init(notifier: AlarmNotifying, soundPlayer: AlarmSoundPlaying) {
        self.notifier = notifier
        self.soundPlayer = soundPlayer
    }

    func handle(decision: GuardDecision, deviceName: String, now: Date) -> AlarmAction {
        if let pausedUntil = pausedUntilByDeviceId[decision.deviceId], pausedUntil > now {
            return .none
        }

        switch decision.alarmIntent {
        case .none:
            stopIfNeeded(deviceId: decision.deviceId)
            return .none
        case .notifyAndVibrate:
            let title = "设备可能离开"
            let body = "\(deviceName) 可能离开看护范围"
            notifier.notify(title: title, body: body, sound: .default)
            return .notify(deviceId: decision.deviceId, title: title, body: body)
        case .ringAndVibrate:
            let title = "设备已离开看护范围"
            let body = "\(deviceName) 仍未恢复连接"
            notifier.notify(title: title, body: body, sound: .defaultCritical)
            activeRingingDeviceIds.insert(decision.deviceId)
            soundPlayer.startRinging()
            return .ring(deviceId: decision.deviceId, title: title, body: body)
        }
    }

    func acknowledge(deviceId: String) -> AlarmAction {
        pausedUntilByDeviceId.removeValue(forKey: deviceId)
        stopIfNeeded(deviceId: deviceId)
        return .stop(deviceId: deviceId)
    }

    func pause(deviceId: String, now: Date, duration: TimeInterval = 600) -> AlarmAction {
        let until = now.addingTimeInterval(duration)
        pausedUntilByDeviceId[deviceId] = until
        stopIfNeeded(deviceId: deviceId)
        return .pause(deviceId: deviceId, until: until)
    }

    func pausedUntil(deviceId: String) -> Date? {
        pausedUntilByDeviceId[deviceId]
    }

    private func stopIfNeeded(deviceId: String) {
        guard activeRingingDeviceIds.remove(deviceId) != nil else { return }
        if activeRingingDeviceIds.isEmpty {
            soundPlayer.stopRinging()
        }
    }
}

final class LocalNotificationAdapter: AlarmNotifying {
    func notify(title: String, body: String, sound: UNNotificationSound?) {
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = sound
        let request = UNNotificationRequest(
            identifier: UUID().uuidString,
            content: content,
            trigger: nil
        )
        UNUserNotificationCenter.current().add(request)
    }
}

final class NoopAlarmSoundPlayer: AlarmSoundPlaying {
    private(set) var isRinging = false

    func startRinging() {
        isRinging = true
    }

    func stopRinging() {
        isRinging = false
    }
}
